lib/parser.rb

```diff
@@ -18,15 +18,14 @@ def advance
     @command = @lines.shift
   end
 
+  NON_ARITHMETIC_COMMANDS = ["push", "pop", "label", "if-goto"]
+
   def command_type
-    if @command.start_with?("push")
-      :C_PUSH
-    elsif @command.start_with?("pop")
-      :C_POP
-    elsif @command.start_with?("label")
-      :C_LABEL
-    elsif @command.start_with?("if-goto")
-      :C_IF
+    raw_command = @command.split[0]
+
+    case raw_command
+    when *NON_ARITHMETIC_COMMANDS
+      "C_#{raw_command.split("-")[0].upcase}".to_sym
```

Comment:
This is clever! It’s certainly concise because it avoids repetition by computing the symbol name directly from the command string.

A potential risk is that this code is _speculatively general_: we’re making assumptions about how all non-arithmetic command strings should be turned into symbols, but we haven’t implemented every command yet, so we don’t necessarily know whether these assumptions will continue to hold in the future. If the remaining VM commands don’t fit this pattern then it might be complicated to deal with any exceptions. (In fact, the unimplemented commands are `function`/`:C_FUNCTION`, `return`/`:C_RETURN` and `call`/`:C_CALL`, which _do_ fit the pattern, so this is more of a hypothetical risk than a real one.)

Another small downside is that clever code is generally harder to understand and maintain. To someone unfamiliar with the project it might not be immediately obvious what this line is doing, and the dynamic symbol generation means that they can no longer easily search for e.g. `:C_LABEL` to find where that symbol gets created.

An alternative approach which makes a different trade-off is to use a hash to store the mapping from command string to symbol:

```ruby
COMMAND_TYPES = {
  'push' => :C_PUSH,
  'pop' => :C_POP,
  'label' => :C_LABEL,
  'if-goto' => :C_IF
}
```

This is more repetitive, but also more explicit, than dynamically generating the symbol. The main advantage is that it’s more straightforwardly flexible in future: if for some reason we are asked to add support for a `foo` command whose type must be `:C_BAR`, this approach makes it trivial to add `'foo' => :C_BAR` to the hash, whereas we’d have to do something more complex in the dynamic version. The explicitness of the hash makes it easy to understand at a glance too.

One incidental advantage is that storing the command types in a hash gives us the opportunity to use the “default if missing” feature of [`Hash#fetch`](https://rubydoc.info/stdlib/core/Hash:fetch):

```ruby
COMMAND_TYPES.fetch(raw_command, :C_ARITHMETIC)
```

This means: if `raw_command` is one of the keys of `COMMAND_TYPES` then return its corresponding value, otherwise return `:C_ARITHMETIC`. So, almost by coincidence, using a hash lets us avoid writing a conditional altogether.